trinity/sync/full/chain.py

```diff
@@ -155,6 +144,87 @@ def __init__(self,
             len(headers), txs, timer.elapsed, head.block_number)
         return head.block_number
 
+    async def _download_receipts(self,
+                                 target_td: int,
+                                 all_headers: Tuple[BlockHeader, ...]) -> None:
+        """
+        Downloads and persists the receipts for the given set of block headers.
+        Receipts are requested from all peers in equal sized batches.
+        """
+        # Post-Byzantium blocks may have identical receipt roots (e.g. when they have the same
+        # number of transactions and all succeed/failed: ropsten blocks 2503212 and 2503284),
+        # so we do this to avoid requesting the same receipts multiple times.
+        headers = tuple(unique(
+            (header for header in all_headers if not _is_receipts_empty(header)),
+            key=operator.attrgetter('receipt_root'),
+        ))
+
+        while headers:
+            # split the remaining headers into equal sized batches for each peer.
+            peers = cast(Tuple[ETHPeer, ...], self.peer_pool.get_peers(target_td))
+            batch_size = math.ceil(len(headers) / len(peers))
+            batches = tuple(partition_all(batch_size, headers))
+
+            # issue requests to all of the peers and wait for all of them to respond.
+            requests = tuple(
+                self._get_receipts(peer, batch)
+                for peer, batch
+                in zip(peers, batches)
+            )
+            responses = await self.wait(asyncio.gather(
+                *requests,
+                loop=self.get_event_loop(),
+            ))
+
+            # extract the returned receipt data and the headers for which we
+            # are still missing receipts.
+            all_receipt_bundles, all_missing_headers = zip(*responses)
+            receipt_bundles = tuple(concat(all_receipt_bundles))
+            headers = tuple(concat(all_missing_headers))
+
+            # process all of the returned receipts, storing their trie data
+            # dicts in the database
+            receipts, trie_roots_and_data_dicts = zip(*receipt_bundles)
+            trie_roots, trie_data_dicts = zip(*trie_roots_and_data_dicts)
+            for trie_data in trie_data_dicts:
+                await self.wait(self.db.coro_persist_trie_data_dict(trie_data))
+        else:
+            self.logger.debug("Got receipts batch for %d headers", len(all_headers))
```

Comment:
This doesn't need to be in an `else` block, right? I did a double-take looking for a `break` statement.